### ISSUE TITLE
spirv-opt: assert/crash with --eliminate-local-multi-store pass

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/SPIRV-Tools#824]
I am using SPIRV-Tools master (to commit c6dfc1188) and SPIRV-Headers master (to commit 2bb92e6fe) and hit an assert/crash with the attached shader in --eliminate-local-multi-store pass.

To repro the assert, you can run the command-line:

spirv-opt.exe --eliminate-local-multi-store lightandcopy.spv -o lightandcopy_opt.spv
Assertion failed: index < operands_.size() && "operand index out of bound", file d:\dev\source2\main\src\thirdparty\spirv-tools\source\opt\instruction.h, line 197

The stack of the assert looks like this:
```
 	vfx_gl.dll!common_assert_to_message_box<wchar_t>(const wchar_t * const expression, const wchar_t * const file_name, const unsigned int line_number, void * const return_address) Line 351	C++
 	vfx_gl.dll!common_assert<wchar_t>(const wchar_t * const expression, const wchar_t * const file_name, const unsigned int line_number, void * const return_address) Line 387	C++
 	vfx_gl.dll!_wassert(const wchar_t * expression, const wchar_t * file_name, unsigned int line_number) Line 405	C++
 	vfx_gl.dll!spvtools::ir::Instruction::GetOperand(unsigned int index) Line 197	C++
 	vfx_gl.dll!spvtools::ir::Instruction::GetSingleWordOperand(unsigned int index) Line 72	C++
 	vfx_gl.dll!spvtools::ir::Instruction::GetSingleWordInOperand(unsigned int index) Line 153	C++
 	vfx_gl.dll!spvtools::opt::LocalMultiStoreElimPass::GetPointeeTypeId(const spvtools::ir::Instruction * ptrInst) Line 176	C++
 	vfx_gl.dll!spvtools::opt::LocalMultiStoreElimPass::PatchPhis(unsigned int header_id, unsigned int back_id) Line 390	C++
>	vfx_gl.dll!spvtools::opt::LocalMultiStoreElimPass::EliminateMultiStoreLocal(spvtools::ir::Function * func) Line 467	C++
 	vfx_gl.dll!spvtools::opt::LocalMultiStoreElimPass::ProcessImpl::__l2::<lambda>(spvtools::ir::Function * fp) Line 550	C++
 	[External Code]	
 	vfx_gl.dll!spvtools::opt::Pass::ProcessEntryPointCallTree(std::function<bool __cdecl(spvtools::ir::Function *)> & pfn, spvtools::ir::Module * module) Line 54	C++
 	vfx_gl.dll!spvtools::opt::LocalMultiStoreElimPass::ProcessImpl() Line 551	C++
 	vfx_gl.dll!spvtools::opt::LocalMultiStoreElimPass::Process(spvtools::ir::Module * module) Line 564	C++
 	vfx_gl.dll!spvtools::opt::PassManager::Run(spvtools::ir::Module * module) Line 23	C++
 	vfx_gl.dll!spvtools::Optimizer::Run(const unsigned int * original_binary, unsigned __int64 original_binary_size, std::vector<unsigned int,std::allocator<unsigned int> > * optimized_binary) Line 77	C++
 	vfx_gl.dll!CVfxShaderValidatorGLslang::GetBinaryShader(unsigned char * * ppBinaryOut, unsigned int * pBinarySizeOut) Line 734	C++
 	vfx_gl.dll!CVfxGL::CompileShaderAndGetVariables(unsigned __int64 nStaticComboIndex, unsigned __int64 nDynamicComboIndex, const CVfx * pVfx, VfxCompileTarget_t compileTarget, VfxProgram_t programType, VfxCompiledShaderInfo_t * pCompiledShaderInfoOut, bool bUseShaderCache, unsigned int nVfxCompileShaderFlags) Line 2860	C++
 	vfxc.exe!ValveMain(bool bIsConsoleApp, void * hInstance, void * hPrevInstance, int nCmdShow) Line 1022	C++
 	vfxc.exe!main(int argc, char * * argv) Line 19	C++
 	[External Code]	
```
We invoke spirv programmatic ally using the recipe below, but I narrowed it down to the CreateLocalMultiStoreElimPass:

```
		optimizer.RegisterPass( spvtools::CreateInlineExhaustivePass() );
		optimizer.RegisterPass( spvtools::CreateLocalAccessChainConvertPass() );
		optimizer.RegisterPass( spvtools::CreateLocalSingleBlockLoadStoreElimPass() );
		optimizer.RegisterPass( spvtools::CreateLocalSingleStoreElimPass() );
		optimizer.RegisterPass( spvtools::CreateInsertExtractElimPass() );
		optimizer.RegisterPass( spvtools::CreateAggressiveDCEPass() );
		optimizer.RegisterPass( spvtools::CreateDeadBranchElimPass() );
		optimizer.RegisterPass( spvtools::CreateBlockMergePass() );
		optimizer.RegisterPass( spvtools::CreateLocalSingleBlockLoadStoreElimPass() );
		optimizer.RegisterPass( spvtools::CreateLocalSingleStoreElimPass() );
		optimizer.RegisterPass( spvtools::CreateLocalMultiStoreElimPass() );
		optimizer.RegisterPass( spvtools::CreateInsertExtractElimPass() );
		optimizer.RegisterPass( spvtools::CreateAggressiveDCEPass() );
		optimizer.RegisterPass( spvtools::CreateCommonUniformElimPass() );
```

[lightandcopy.zip](https://github.com/KhronosGroup/SPIRV-Tools/files/1314126/lightandcopy.zip)

**Comment by dneto0** [comment]
Here's my trace with debug info:

#0  0x00007ffff743dc37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff7441028 in __GI_abort () at abort.c:89
#2  0x00007ffff7436bf6 in __assert_fail_base (fmt=0x7ffff758b018 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x7af578 "index < operands_.size() && \"operand index out of bound\"", file=file@entry=0x7af528 "/build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/instruction.h", 
    line=line@entry=197, 
    function=function@entry=0x7af680 <spvtools::ir::Instruction::GetOperand(unsigned int) const::__PRETTY_FUNCTION__> "const spvtools::ir::Operand& spvtools::ir::Instruction::GetOperand(uint32_t) const")
    at assert.c:92
#3  0x00007ffff7436ca2 in __GI___assert_fail (assertion=0x7af578 "index < operands_.size() && \"operand index out of bound\"", 
    file=0x7af528 "/build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/instruction.h", line=197, 
    function=0x7af680 <spvtools::ir::Instruction::GetOperand(unsigned int) const::__PRETTY_FUNCTION__> "const spvtools::ir::Operand& spvtools::ir::Instruction::GetOperand(uint32_t) const")
    at assert.c:101
#4  0x00000000006603eb in spvtools::ir::Instruction::GetOperand (this=0xac9270, index=2) at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/instruction.h:197
#5  0x00000000006c07ee in spvtools::ir::Instruction::GetSingleWordOperand (this=0xac9270, index=2) at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/instruction.cpp:72
#6  0x0000000000660378 in spvtools::ir::Instruction::GetSingleWordInOperand (this=0xac9270, index=1) at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/instruction.h:152
#7  0x00000000006d9781 in spvtools::opt::LocalMultiStoreElimPass::GetPointeeTypeId (this=0xa99d00, ptrInst=0xafe510)
    at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/local_ssa_elim_pass.cpp:175
#8  0x00000000006db023 in spvtools::opt::LocalMultiStoreElimPass::PatchPhis (this=0xa99d00, header_id=13298, back_id=22301)
    at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/local_ssa_elim_pass.cpp:392
#9  0x00000000006db6bb in spvtools::opt::LocalMultiStoreElimPass::EliminateMultiStoreLocal (this=0xa99d00, func=0xadc430)
    at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/local_ssa_elim_pass.cpp:466
#10 0x00000000006dbcde in spvtools::opt::LocalMultiStoreElimPass::__lambda13::operator() (__closure=0xafd450, fp=0xadc430)
    at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/local_ssa_elim_pass.cpp:550
#11 0x00000000006dd51e in std::_Function_handler<bool(spvtools::ir::Function*), spvtools::opt::LocalMultiStoreElimPass::ProcessImpl()::__lambda13>::_M_invoke(const std::_Any_data &, spvtools::ir::Function *) (__functor=..., __args#0=0xadc430) at /usr/include/c++/4.8/functional:2057
#12 0x00000000006e5e85 in std::function<bool (spvtools::ir::Function*)>::operator()(spvtools::ir::Function*) const (this=0x7fffffffd090, __args#0=0xadc430) at /usr/include/c++/4.8/functional:2471
#13 0x00000000006e5b16 in spvtools::opt::Pass::ProcessEntryPointCallTree(std::function<bool (spvtools::ir::Function*)>&, spvtools::ir::Module*) (this=0xa99d00, pfn=..., module=0xab9320)
    at /build/dneto/v/vokuhila/third_party/spirv-tools/source/opt/pass.cpp:54

**Comment by dneto0** [comment]
I have a stack trace with debug info:
#0  0x00007ffff743dc37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff7441028 in __GI_abort () at abort.c:89
#2  0x00007ffff7436bf6 in __assert_fail_base (fmt=0x7ffff758b018 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x7af578 "index < operands_.size() && \"operand index out of bound\"", file=file@entry=0x7af528 "//spirv-tools/source/opt/instruction.h", 
    line=line@entry=197, 
    function=function@entry=0x7af680 <spvtools::ir::Instruction::GetOperand(unsigned int) const::__PRETTY_FUNCTION__> "const spvtools::ir::Operand& spvtools::ir::Instruction::GetOperand(uint32_t) const")
    at assert.c:92
#3  0x00007ffff7436ca2 in __GI___assert_fail (assertion=0x7af578 "index < operands_.size() && \"operand index out of bound\"", 
    file=0x7af528 "//spirv-tools/source/opt/instruction.h", line=197, 
    function=0x7af680 <spvtools::ir::Instruction::GetOperand(unsigned int) const::__PRETTY_FUNCTION__> "const spvtools::ir::Operand& spvtools::ir::Instruction::GetOperand(uint32_t) const")
    at assert.c:101
#4  0x00000000006603eb in spvtools::ir::Instruction::GetOperand (this=0xac9270, index=2) at //spirv-tools/source/opt/instruction.h:197
#5  0x00000000006c07ee in spvtools::ir::Instruction::GetSingleWordOperand (this=0xac9270, index=2) at //spirv-tools/source/opt/instruction.cpp:72
#6  0x0000000000660378 in spvtools::ir::Instruction::GetSingleWordInOperand (this=0xac9270, index=1) at //spirv-tools/source/opt/instruction.h:152
#7  0x00000000006d9781 in spvtools::opt::LocalMultiStoreElimPass::GetPointeeTypeId (this=0xa99d00, ptrInst=0xafe510)
    at //spirv-tools/source/opt/local_ssa_elim_pass.cpp:175
#8  0x00000000006db023 in spvtools::opt::LocalMultiStoreElimPass::PatchPhis (this=0xa99d00, header_id=13298, back_id=22301)
    at //spirv-tools/source/opt/local_ssa_elim_pass.cpp:392
#9  0x00000000006db6bb in spvtools::opt::LocalMultiStoreElimPass::EliminateMultiStoreLocal (this=0xa99d00, func=0xadc430)
    at //spirv-tools/source/opt/local_ssa_elim_pass.cpp:466
#10 0x00000000006dbcde in spvtools::opt::LocalMultiStoreElimPass::__lambda13::operator() (__closure=0xafd450, fp=0xadc430)
    at //spirv-tools/source/opt/local_ssa_elim_pass.cpp:550
#11 0x00000000006dd51e in std::_Function_handler<bool(spvtools::ir::Function*), spvtools::opt::LocalMultiStoreElimPass::ProcessImpl()::__lambda13>::_M_invoke(const std::_Any_data &, spvtools::ir::Function *) (__functor=..., __args#0=0xadc430) at /usr/include/c++/4.8/functional:2057
#12 0x00000000006e5e85 in std::function<bool (spvtools::ir::Function*)>::operator()(spvtools::ir::Function*) const (this=0x7fffffffd090, __args#0=0xadc430) at /usr/include/c++/4.8/functional:2471
#13 0x00000000006e5b16 in spvtools::opt::Pass::ProcessEntryPointCallTree(std::function<bool (spvtools::ir::Function*)>&, spvtools::ir::Module*) (this=0xa99d00, pfn=..., module=0xab9320)
    at //spirv-tools/source/opt/pass.cpp:54
#14 0x00000000006dbe65 in spvtools::opt::LocalMultiStoreElimPass::ProcessImpl (this=0xa99d00) at //spirv-tools/source/opt/local_ssa_elim_pass.cpp:551
#15 0x00000000006dc1c7 in spvtools::opt::LocalMultiStoreElimPass::Process (this=0xa99d00, module=0xab9320) at //spirv-tools/source/opt/local_ssa_elim_pass.cpp:563
#16 0x000000000067347b in spvtools::opt::PassManager::Run (this=0xa99c98, module=0xab9320) at //spirv-tools/source/opt/pass_manager.cpp:23
#17 0x00000000006671e0 in spvtools::Optimizer::Run (this=0x7fffffffd220, original_binary=0xa9d6b0, original_binary_size=2904, optimized_binary=0x7fffffffd260)

**Comment by greg-lunarg** [comment]
I have recreated this with VS.

**Comment by dneto0** [comment]
I'm trying to reduce the test case.

**Comment by greg-lunarg** [comment]
One unusual aspect to this shader is that even with exhaustive inlining, at least function is not inlining into the entry point. The failure is occurring in the called function. This is likely related to the problem as shaders with non-inlined calls have not gotten much exercise.

**Comment by greg-lunarg** [comment]
Fix submitted PR #825

**Comment by dneto0** [comment]
Something fishy is going on.  We have multiple functions F1 and F2.  When traversing blocks in structured order in F2, we end up seeing a block from F1. Strangely.

And.... I'll look at Greg's patch.  :-)

**Comment by dneto0** [comment]
Fixed by #825
